### DISCUSSION TITLE
Calculation of porosity in a part of the network

### DISCUSSION THREAD

**Discussion by Devin-one** [PMEAL/OpenPNM#2556] [Q&A]
Dear Professor, what should I do if I want to calculate the porosity of a part of the network? For example, calculate the porosity of the `pn.pores('left')` and the `pn.find_neighbor_throats`.

**Comment by jgostick** [comment]
Hi @Liqiwen123...I changed my email setup and have not been getting notifications from github, sorry the late respone!

RE: your question...this is a bit tricky.  Porosity in general is a bit tricky to calculate.  Is the network cubic?  If so then you could use geometrical values to find it.  As in: compute the pore and throat volume of the pores of interest and divide by the volume of the lattice cells in which the pores lie. If it's random, then the only option is to generate an image, then just use image analysis.  There is a function in porespy.networks for this called `generate_voxel_image`.  Once you have the image, you can just isolate the section you want then compute it's porosity as ``im.sum()/im.size``.

**Reply by Devin-one** [reply]
Yes, this network is cubic, `pn = op.network.Cubic(shape=[30,20,1], spacing=1)`.  I only know the code for calculating the pore volume of the whole network, as `np.sum(pn['pore.volume'])`.  However, I do not know how to calculate the pore volume of the pores I am interested in. For example, if I want to calculate the pore volume of `pores('left')`, can you tell me what code to use? Thank you in advance.

**Reply by jgostick** [reply]
The hardest part will be finding the throat volume, since these will intersect the boundary of your 'sub-domain'.  We don't have a way to do this, but it's not impossible.  You can identify the pores of interest (i.e. `Ps = pn.pores('left')`), then find their volume (`Vp = pn['pore.volume'][Ps].sum()`).  Then you'll want to find all the neighboring throats that are shared (`Ts = pn.find_neighbor_throats(pores=Ps, mode='xnor'`) and find their volume (`Vt1 = pn['throat.volume'][Ts].sum()`).  Lastly, you'll want to find the throats that are connected to pores that are outside your group of interest (`Ts = pn.find_neigbor_pores(pores=Ps, mode='xor'`).  This is the tricky part. As a *first approximation* you could final their volume, then divided it by 2, assuming half of the volume is within your domain of interest, so (`Vt2 = pn['throat.volume'][Ts].sum()/2`).  Doing it the fully rigorous way, by finding the exact amount of volume that lies within the domain of interest, is probably not going to change the answer very much.

> Note: If you are assuming sphereical pores the you need to account for the overlap between the pore bodies and throats.  There is a tutorial which is for V3, but talks about this [here](https://openpnm.org/examples/tutorials/05_defining_geometry.html)

**Reply by xiangWu-WW** [reply]
> Yes, this network is cubic, `pn = op.network.Cubic(shape=[30,20,1], spacing=1)`. I only know the code for calculating the pore volume of the whole network, as `np.sum(pn['pore.volume'])`. However, I do not know how to calculate the pore volume of the pores I am interested in. For example, if I want to calculate the pore volume of `pores('left')`, can you tell me what code to use? Thank you in advance.

For the porosity,i think you should consider the volume of throats.
